### PR TITLE
Fix generated JSX for custom elements

### PR DESCRIPTION
**Pull request by mdynnl**
Related-to: https://github.com/remarkjs/remark-math/issues/72 https://github.com/wooorm/xdm/commit/ddcebdbcb577015a8930b310ee7a3ba045f1c24a

### CODE REVIEW

File: lib/plugin/recma-jsx-rewrite.js [hunk 1]
```diff
@@ -182,10 +179,7 @@ export function recmaJsxRewrite(options = {}) {
             }
 
             if (node.closingElement) {
-              node.closingElement.name = toJsxIdOrMemberExpression([
-                '_components',
-                id
-              ])
+              node.closingElement.name = node.openingElement.name
```
Comment (wooorm) [inline lib/plugin/recma-jsx-rewrite.js:182]: We can’t reuse this object by reference. It has to be a new node unfortunately

**Review by wooorm — COMMENTED** [review]
Hey! Thanks for working on a PR!

I think a better solution is likely to fix the faulty function rather than to stop using the problematic code here.

The code in `estree-util-to-id-or-member-expressions.js` attempts to support JSX and ES identifiers. It exports methods for both. Those methods use the correct labels, but they should also behave differently.

Here is some code that is similar to the dependency `estree-util-is-identifier-name`, so it checks whether some value is an identifier, but it checks whether a value is a JSX identifier (it allows `-`, when not at the start).

```js
import {name as isIdentifierName, start, cont as esCont} from 'estree-util-is-identifier-name'

// …

/**
 * Checks if the given character code can continue a JSX identifier.
 * @param {number} code
 */
function cont(code) {
  return code === 45 /* `-` */ || start(code) || contRe.test(String.fromCharCode(code))
}

/**
 * Checks if the given string is a valid JSX identifier name.
 * @param {string} name
 */
function isJsxIdentifierName(name) {
  let index = -1

  while (++index < name.length) {
    if (!(index ? cont : start)(name.charCodeAt(index))) return false
  }

  // `false` if `name` is empty.
  return index > 0
}
```

This code can be placed in `estree-util-to-id-or-member-expression.js`.
The function call to create `toIdOrMemberExpression` can pass the normal `isIdentifierName` and the function call for `toJsxIdOrMemberExpression` can pass the new `isJsxIdentifierName` to the function `toIdOrMemberExpressionFactory`, which than uses the dynamic passed function instead of the static `isIdentifierName` (L41).

I believe that should do the trick?

File: lib/plugin/recma-jsx-rewrite.js (hunk 1 above)
Comment (mdynnl) [reply lib/plugin/recma-jsx-rewrite.js:182]: Sorry for my lack of understanding. Didn't think that would be problematic.

**Comment by mdynnl** [comment]
Yes, thought about a new function that would support jsx ids, but not confident enough to do that. I'll try again with the above code.

**Comment by wooorm** [comment]
Does that work? Need help?

File: lib/plugin/recma-jsx-rewrite.js (hunk 1 above)
Comment (wooorm) [reply lib/plugin/recma-jsx-rewrite.js:182]: No problem at all! 🙂

**Comment by mdynnl** [comment]
Yes, worked after a little fix `contRe.test(String.fromCharCode(code))` to `esCont(code)`. Is that correct? But something came up and didn't push. I'm back now. Will push if the above is correct.

**Comment by mdynnl** [comment]
Or maybe it also works if `contRe` is imported from esutil.

**Comment by wooorm** [comment]
Ah yes I made a mistake there. I meant cont, and calling it!

**Comment by mdynnl** [comment]
I'm behind a proxy and can't run the full test. Only ran `test/core`.

**Comment by mdynnl** [comment]
> Ah yes I made a mistake there. I meant cont, and calling it!

And I thought it was intentional and a test to me ^_^

**Review by wooorm — COMMENTED** [review]
Thanks for your continued work on this. As you can see from the tests, it doesn’t work yet, so there are some more things to do.

Please see [the last paragraph of my previous review](https://github.com/wooorm/xdm/pull/106#pullrequestreview-853674593). Note that this file exports **two functions**. One is about JSX identifiers, which **must use `isJsxIdentifierName`**. The other is about normal identifiers, which **must use `isIdentifierName`**.

You can solve this by passing those functions at line 13/14 (`isIdentifierName`) and the other on L19 (`isJsxIdentifierName`).
Then you can receive that function on L25 as a parameter, and use it on L41!

**Comment by mdynnl** [comment]
Thanks for pointing that out again. I got it but can't meet the coverage for branches. Still at 99.75%.

**Comment by mdynnl** [comment]
c8 is right. `JSXMemberExpression` has no child node of `Literal` type. That branch will never execute given the amount of the tests written including the one I added. I wonder if we still need `toJsxIdOrMemberExpression`. Instead construct the node manually like `openingElement` and it worked fine until now.

**Comment by mdynnl** [comment]
This was changed from manual node construction to function in https://github.com/mdynnl/xdm/commit/62e6f30af2fb4d2fca73c55ca8d605615c6d27cb

Reverting that change would still work fine with dev errors, isn't it?

**Review by wooorm — APPROVED** [review]

**Comment by mdynnl** [comment]
Oh, should have pushed anyway. I was waiting for your reply. Thanks for dealing with this.

**Comment by wooorm** [comment]
Thanks Nyi Nyi. I implemented the passing and receiving of identifier checks.
It is indeed impossible to hit 100% coverage normally. That is okay.
It is an edge case that shouldn’t occur, similar to edge case we are currently solving.
By throwing an error there we can ensure that in the future, if something weird gets passed, we can solve it earlier due to the error message.

**Comment by mdynnl** [comment]
Sorry if I caused more problems rather than solved. First time contribution. But I learnt something. So, thanks.

**Comment by wooorm** [comment]
Released, thanks for your work!

No problem. That’s only natural. The second time will be easier, and so on! ;)

**Pull request merged**